libsolidity/ast/Types.cpp

```diff
@@ -362,6 +369,74 @@ MemberList::MemberMap Type::boundFunctions(Type const& _type, ASTNode const& _sc
 	if (auto refType = dynamic_cast<ReferenceType const*>(&_type))
 		typeLocation = refType->location();
 
+	return usingForDirectives | ranges::views::filter([&](UsingForDirective const* _directive) -> bool {
+		// Convert both types to pointers for comparison to see if the `using for` directive applies.
+		// Note that at this point we don't yet know if the functions are actually usable with the type.
+		// `_type` may not be convertible to the function parameter type.
+		return
+			!_directive->typeName() ||
+			*TypeProvider::withLocationIfReference(typeLocation, &_type, true) ==
+			*TypeProvider::withLocationIfReference(
+				typeLocation,
+				_directive->typeName()->annotation().type,
+				true
+			);
+	}) | ranges::to<vector<UsingForDirective const*>>;
+}
+
+}
+
+Result<FunctionDefinition const*> Type::userDefinedOperator(Token _token, ASTNode const& _scope, bool _unaryOperation) const
+{
+	if (!typeDefinition())
+		return nullptr;
+
+	set<FunctionDefinition const*> matchingDefinitions;
+	for (UsingForDirective const* ufd: usingForDirectivesForType(*this, _scope))
+		for (auto const& [identifierPath, operator_]: ufd->functionsAndOperators())
+		{
+			if (operator_ != _token)
+				continue;
+			FunctionDefinition const& function = dynamic_cast<FunctionDefinition const&>(
+				*identifierPath->annotation().referencedDeclaration
+			);
+			FunctionType const* functionType = dynamic_cast<FunctionType const*>(
+				function.libraryFunction() ? function.typeViaContractName() : function.type()
+			);
+			solAssert(functionType && !functionType->parameterTypes().empty());
+
+			Type const* normalizedType = this;
+			if (auto const* referenceType = dynamic_cast<ReferenceType const*>(normalizedType))
+				normalizedType = TypeProvider::withLocationIfReference(referenceType->location(), normalizedType);
+
+			Type const* normalizedParameterType = functionType->parameterTypes().front();
+			if (auto const* referenceType = dynamic_cast<ReferenceType const*>(normalizedParameterType))
+				normalizedParameterType = TypeProvider::withLocationIfReference(referenceType->location(), normalizedParameterType);
+
+			if (
+				*normalizedType == *normalizedParameterType &&
+				(
+					(_unaryOperation && function.parameterList().parameters().size() == 1) ||
+					(!_unaryOperation && function.parameterList().parameters().size() == 2)
+				)
+			)
+				matchingDefinitions.insert(&function);
+		}
+
+	if (matchingDefinitions.size() == 1)
+		return *matchingDefinitions.begin();
+	else if (util::contains(util::overridableOperators, _token))
```

Comment:
Why not check this condition before we do all the hard work? :) It does not depend on what functions we find. You could put it at the beginning of the function, next to `if (!typeDefinition())`.

Reply:
I meant that parser has his check:
```c++
			if (!util::contains(overridableOperators, operator_))
```
but I now realized that parser can do this only for the `using for` directive and here `_token` is an operator we encountered in an expression.

Still, I'd make it an assertion and have the caller check this.